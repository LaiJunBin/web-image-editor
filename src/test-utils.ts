const fn = vi.fn()
export type fn = typeof fn
