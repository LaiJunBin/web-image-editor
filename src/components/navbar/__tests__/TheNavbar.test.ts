import type { MenuType } from '@/const/menus'
import { shallowMount } from '@vue/test-utils'
import TheNavbarVue from '../TheNavbar.vue'
import { nextTick } from 'vue'

enum TestMenuAction {
  CHILD1,
  CHILD2,
  CHILD3
}

const menus: MenuType<TestMenuAction>[] = [
  {
    text: 'Parent 1',
    children: [
      {
        text: 'Child 1',
        action: TestMenuAction.CHILD1
      },
      {
        text: 'Child 2',
        action: TestMenuAction.CHILD2
      }
    ]
  },
  {
    text: 'Parent 2',
    children: [
      {
        text: 'Child 3',
        action: TestMenuAction.CHILD3
      }
    ]
  }
]

describe('TheNavbar', () => {
  test('render properly menus', async () => {
    const wrapper = shallowMount(TheNavbarVue<TestMenuAction>, {
      attachTo: document.body,
      props: {
        menus
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('open/close menu', async () => {
    const wrapper = shallowMount(TheNavbarVue<TestMenuAction>, {
      attachTo: document.body,
      props: {
        menus
      }
    })

    const targetMenu = wrapper.findAll('li').find((li) => li.text().match(menus[0].text))
    expect(targetMenu).toBeDefined()

    const targetSubMenu = targetMenu!.get('ul')
    expect(targetSubMenu.isVisible()).toBeFalsy()

    await targetMenu?.trigger('click')
    expect(targetSubMenu.isVisible()).toBeTruthy()

    await targetMenu?.trigger('click')
    expect(targetSubMenu.isVisible()).toBeFalsy()
  })

  test('emit menu action', async () => {
    const wrapper = shallowMount(TheNavbarVue<TestMenuAction>, {
      attachTo: document.body,
      props: {
        menus
      }
    })

    const targetMenu = wrapper.findAll('li').find((li) => li.text().match(menus[0].text))
    expect(targetMenu).toBeDefined()

    const targetItem = targetMenu!
      .findAll('li')
      .find((li) => li.text().match(menus[0].children[0].text))
    expect(targetItem).toBeDefined()

    await targetItem?.trigger('click')
    expect(wrapper.emitted('menu-action')).toEqual([[menus[0].children[0].action]])
  })

  test('close menu when click outside', async () => {
    const wrapper = shallowMount(TheNavbarVue<TestMenuAction>, {
      attachTo: document.body,
      props: {
        menus
      }
    })

    const targetMenu = wrapper.findAll('li').find((li) => li.text().match(menus[0].text))
    expect(targetMenu).toBeDefined()

    const targetSubMenu = targetMenu!.get('ul')
    expect(targetSubMenu.isVisible()).toBeFalsy()

    await targetMenu?.trigger('click')
    expect(targetSubMenu.isVisible()).toBeTruthy()

    window.dispatchEvent(new Event('click'))
    await nextTick()
    expect(targetSubMenu.isVisible()).toBeFalsy()
  })

  test('unmount', async () => {
    const wrapper = shallowMount(TheNavbarVue<TestMenuAction>, {
      attachTo: document.body,
      props: {
        menus
      }
    })

    const targetMenu = wrapper.findAll('li').find((li) => li.text().match(menus[0].text))
    expect(targetMenu).toBeDefined()

    const targetSubMenu = targetMenu!.get('ul')
    expect(targetSubMenu.isVisible()).toBeFalsy()

    await targetMenu?.trigger('click')
    expect(targetSubMenu.isVisible()).toBeTruthy()

    wrapper.unmount()
    window.dispatchEvent(new Event('click'))
    await nextTick()
    expect(targetSubMenu.isVisible()).toBeTruthy()
  })
})
