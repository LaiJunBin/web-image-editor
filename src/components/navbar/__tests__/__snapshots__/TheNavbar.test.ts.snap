// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TheNavbar > render properly menus 1`] = `
"<nav class=\\"z-50 select-none bg-neutral-700\\">
  <ul class=\\"flex items-center justify-start\\">
    <li class=\\"relative cursor-pointer px-2 py-1 transition hover:bg-zinc-600\\"><span>Parent 1</span>
      <ul class=\\"absolute left-0 top-full w-max bg-neutral-500\\" style=\\"display: none;\\">
        <li class=\\"px-2 py-1 transition hover:bg-neutral-400\\"><span>Child 1</span></li>
        <li class=\\"px-2 py-1 transition hover:bg-neutral-400\\"><span>Child 2</span></li>
      </ul>
    </li>
    <li class=\\"relative cursor-pointer px-2 py-1 transition hover:bg-zinc-600\\"><span>Parent 2</span>
      <ul class=\\"absolute left-0 top-full w-max bg-neutral-500\\" style=\\"display: none;\\">
        <li class=\\"px-2 py-1 transition hover:bg-neutral-400\\"><span>Child 3</span></li>
      </ul>
    </li>
  </ul>
</nav>"
`;
